File: lib/utils/migrate_assets.dart
import 'dart:io';
import 'dart:convert';

void migrateComponent(Map<String, dynamic> comp) {
  if (comp.containsKey('blueprint')) {
    var bp = comp['blueprint'] as Map<String, dynamic>;
    comp['internalComponents'] = bp['components'] ?? [];
    comp['internalConnections'] = bp['connections'] ?? [];
    comp['name'] = bp['name'] ?? comp['name'] ?? 'IC';
    comp.remove('blueprint');
    
    // Recursively migrate nested components
    var internals = comp['internalComponents'] as List;
    for (var child in internals) {
      if (child is Map<String, dynamic>) {
        migrateComponent(child);
      }
    }
  }
}

Future<void> migrateFile(File file) async {
  print('Migrating ${file.path}...');
  try {
    String content = await file.readAsString();
    var data = jsonDecode(content);
    
    if (data is Map<String, dynamic> && data.containsKey('components')) {
      var comps = data['components'] as List;
      for (var comp in comps) {
        if (comp is Map<String, dynamic>) {
          migrateComponent(comp);
        }
      }
    }
    
    await file.writeAsString(JsonEncoder.withIndent('  ').convert(data));
  } catch (e) {
    print('Error migrating ${file.path}: $e');
  }
}

void main() async {
  var assetsDir = Directory(r'c:\Users\Robert\Documents\digital_bricks\assets');
  if (!await assetsDir.exists()) {
    print('Assets directory not found!');
    return;
  }
  
  await for (var entity in assetsDir.list()) {
    if (entity is File && entity.path.endsWith('.json')) {
      await migrateFile(entity);
    }
  }
  print('Migration complete.');
}

File: python_generators/migrate_assets.py
import json
import os

def migrate_component(comp):
    if 'blueprint' in comp:
        bp = comp['blueprint']
        comp['internalComponents'] = bp.get('components', [])
        comp['internalConnections'] = bp.get('connections', [])
        comp['name'] = bp.get('name', comp.get('name', 'IC'))
        del comp['blueprint']
        
        # Recursively migrate nested components
        for child in comp['internalComponents']:
            migrate_component(child)

def migrate_file(filepath):
    print(f"Migrating {filepath}...")
    with open(filepath, 'r', encoding='utf-8') as f:
        try:
            data = json.load(f)
        except json.JSONDecodeError as e:
            print(f"Error decoding {filepath}: {e}")
            return

    # Check root level components
    if 'components' in data:
        for comp in data['components']:
            migrate_component(comp)
            
    with open(filepath, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2)

def main():
    assets_dir = r"c:\Users\Robert\Documents\digital_bricks\assets"
    for filename in os.listdir(assets_dir):
        if filename.endswith(".json"):
            filepath = os.path.join(assets_dir, filename)
            # We already know which files have "blueprint" from rg, but let's process all for safety
            migrate_file(filepath)

if __name__ == "__main__":
    main()
